pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -204,10 +205,12 @@ public long getLastSequenceId() {
     }
 
     @Override
-    CompletableFuture<MessageId> internalSendAsync(Message<T> message) {
-        CompletableFuture<MessageId> future = new CompletableFuture<>();
+    CompletableFuture<MessageId> internalSendAsync(Message<T> message) {CompletableFuture<MessageId> future = new CompletableFuture<>();
 
-        sendAsync(message, new SendCallback() {
+        MessageImpl<T> interceptorMessage = (MessageImpl<T>) beforeSend(message);
+        interceptorMessage.getDataBuffer().retain();
```

Comment:
Can you add a comment on why we are retaining the buffer here and when it will be released?

Reply:
ok